### ISSUE TITLE
PyTorch build from source with pybind-dev system package installed crashes with `cannot initialize type "Options": an object with that name is already defined` after import

### ISSUE THREAD

**Issue by Algomorph** [pytorch/pytorch#58750]
## 🐛 Bug

I'm getting an error trying to import pytorch.

## To Reproduce
After installing v1.9.0-rc1 via `python setup.py bdist_wheel` and `pip install ./dist/torch-1.9.0a0+gitc94e689-cp38-cp38-linux_x86_64.whl`, when I `import torch` I get the following error:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  generic_type: cannot initialize type "Options": an object with that name is already defined
Aborted (core dumped)
````

after which, the kernel dies. I'm still not sure if it's a PyTorch bug or something related to my specific environment. 

## Environment

Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.18.4
Libc version: glibc-2.31

Python version: 3.8 (64-bit runtime)
Python platform: Linux-5.4.0-73-generic-x86_64-with-glibc2.29
Is CUDA available: N/A
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1070
Nvidia driver version: 460.56
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.2.0
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.20.2
[pip3] pytorch3d==0.4.0
[pip3] torch==1.9.0a0+git145a6f7
[conda] Could not collect

 - How you installed PyTorch (`conda`, `pip`, source): source & pip
 - Build command you used (if compiling from source): `python setup.py bdist_wheel`
 - Python version: 3.8.5 
 - CUDA/cuDNN version: ~11.2~ 11.1 / 8.2.0
 - GPU models and configuration: (see above output)
 - Any other relevant information: conda not installed, no env used, installing for user only

## Other Context Information

The wheels for 1.8.1 and 1.7.1 seem to work fine/don't suffer from this issue.

I really think I need one of the fixes that was applied to torch after the 1.8.1 release, specifically, this one: https://github.com/pytorch/pytorch/pull/57323. I tried making a patch, but it seems like just patching 1.8.1 with that one commit won't work (possibly, because there were other changes to the same files since 1.8.1?). If there is a way to apply the changes over 1.8.1, that could possibly work for me.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @malfet @seemethere @walterddr

**Comment by malfet** [comment]
@Algomorph can you please share few more details about your system: 
 - What CUDA version do you have and how you've installed it (looks like our automated script fails to detect it)
 - Do you have ld.gold installed on your machine by any change?
 - If so, can you please try recompiling pytorch with USE_KINETO=OFF option
 - Does the crash reproduces if you are to use nightly or 1.9.0 RC1 build from [here](https://download.pytorch.org/whl/test/cu102/torch-1.9.0%2Bcu102-cp38-cp38-linux_x86_64.whl)

**Comment by seemethere** [comment]
Can you check if this is also the case on master as well?

**Comment by Algomorph** [comment]
@seemethere Yep, it is also the case on the master, already checked.
@malfet:

- CUDA version: 11.2  (CuDNN version: 8.2.0), as per the original post
- ld.gold not installed
- Just tried recompiling with USE_KINETO=OFF, same bug occurs during import
- The bug does not recur if I try to use the 1.9.0 RC1 build from [here](https://download.pytorch.org/whl/test/cu102/torch-1.9.0%2Bcu102-cp38-cp38-linux_x86_64.whl)

**Comment by Algomorph** [comment]
P.S. correction: the CUDA being used to build everything is 11.1, not 11.2. My mistake.

**Comment by Algomorph** [comment]
Gentlemen, happy to say--I found the culprit for this bug. It was the pybind11 version. Thank you for looking into this regardless. The stock pybind11 version for the standard [pybind11-dev](https://packages.ubuntu.com/focal/pybind11-dev) package for Ubuntu is 2.4.3.

Workaround: remove the pybind11-dev package via apt, build pybind11 2.6.2 from source and install in `usr/local`, then repeat the steps for building the pip wheel for pytorch.

What baffles me is why, when there is a correct version checked out locally (via submodules) in `torch_repo_root/third_party/pybind11`, does PyTorch's CMake still go out looking for a system-installed pybind11? I've tried to analyze `torch_repo_root/cmake/Modules/Findpybind11.cmake`. Perhaps the `find_path` command in the following segment could use some hints to search in the local submodule folder first, instead of the system include directory?

https://github.com/pytorch/pytorch/blob/5c7dace309cc84cc17629172ce97566285a9da58/cmake/Modules/Findpybind11.cmake#L5-L8

**Comment by Algomorph** [comment]
Upon further investigation, it seems like this mechanism in Dependencies.cmake is not sensitive to pybind11 version requirements:
https://github.com/pytorch/pytorch/blob/5c7dace309cc84cc17629172ce97566285a9da58/cmake/Dependencies.cmake#L1002-L1022

I tried to fiddle with it, e.g. by adding the version 2.6.2 to the arguments of both `find_package` calls, but that doesn't seem to do the trick... I still get this output:

```
CMake Warning at cmake/Dependencies.cmake:1003 (find_package):
  Could not find a configuration file for package "pybind11" that is
  compatible with requested version "2.6.2".

  The following configuration files were considered but not accepted:

    /usr/lib/cmake/pybind11/pybind11Config.cmake, version: 2.4.3

Call Stack (most recent call first):
  CMakeLists.txt:617 (include)


-- Found pybind11: /usr/include (Required is at least version "2.6.2") 
-- System pybind11 found
-- pybind11 include dirs: /usr/include
```

**Comment by malfet** [comment]
@Algomorph thank you for the thorough investigation. Going to change labels and title to reflect what is going on.

**Comment by adamjstewart** [comment]
I had this same problem with pybind11 2.5.0, it would be nice if PyTorch could keep track of which versions of dependencies work and check for this during cmake configuration.

**Comment by Adam27X** [comment]
I also ran into this issue when installing pytorch 1.9.1.
